rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -148,56 +156,77 @@ object Reduce {
       * @param par
       * @return
       */
-    override def eval(par: Par)(implicit env: Env[Par]): M[Unit] = {
-      def handle[A](eval: (A => M[Unit]))(a: A): M[Unit] =
-        eval(a).handleError((e: InterpreterError) => {
+    override def eval(par: Par)(implicit env: Env[Par], rand: Blake2b512Random): M[Unit] = {
+      val filteredExprs = par.exprs.filter { expr =>
+        expr.exprInstance match {
+          case _: EVarBody    => true
+          case _: EEvalBody   => true
+          case _: EMethodBody => true
+          case _              => false
+        }
+      }
+      val starts = Vector(par.sends.size,
+                          par.receives.size,
+                          par.news.size,
+                          par.matches.size,
+                          par.bundles.size,
+                          filteredExprs.size)
+        .scanLeft(0)(_ + _)
+      def handle[A](eval: (A => (Env[Par], Blake2b512Random) => M[Unit]), start: Int)(
+          ta: (A, Int)): M[Unit] = {
+        val newRand =
+          if (starts(6) == 1)
+            rand
+          else if (starts(6) > 256)
+            rand.splitShort((start + ta._2).toShort)
+          else
+            rand.splitByte((start + ta._2).toByte)
+        eval(ta._1)(env, newRand).handleError((e: InterpreterError) => {
           fTell.tell(e)
         })
+      }
       List(
-        Parallel.parTraverse(par.sends.toList)(handle(eval)),
-        Parallel.parTraverse(par.receives.toList)(handle(eval)),
-        Parallel.parTraverse(par.news.toList)(handle(eval)),
-        Parallel.parTraverse(par.matches.toList)(handle(eval)),
-        Parallel.parTraverse(par.bundles.toList)(handle(eval)),
-        Parallel.parTraverse(par.exprs.filter { expr =>
-          expr.exprInstance match {
-            case _: EVarBody    => true
-            case _: EEvalBody   => true
-            case _: EMethodBody => true
-            case _              => false
-          }
-        }.toList)(expr =>
+        Parallel.parTraverse(par.sends.zipWithIndex.toList)(handle(evalExplicit, starts(0))),
+        Parallel.parTraverse(par.receives.zipWithIndex.toList)(handle(evalExplicit, starts(1))),
+        Parallel.parTraverse(par.news.zipWithIndex.toList)(handle(evalExplicit, starts(2))),
+        Parallel.parTraverse(par.matches.zipWithIndex.toList)(handle(evalExplicit, starts(3))),
+        Parallel.parTraverse(par.bundles.zipWithIndex.toList)(handle(evalExplicit, starts(4))),
+        Parallel.parTraverse(filteredExprs.zipWithIndex.toList)(texpr => {
+          val (expr, idx) = texpr
+          val newRand =
+            if (starts(6) == 1)
+              rand
+            else if (starts(6) > 256)
+              rand.splitShort((starts(5) + idx).toShort)
```

Comment:
I don't get it why you first check number of expressions (`starts(6)`) but you split on the bundles? I'm lost.

Reply:
Starts(0) is always 0, 
Starts(1) is sends.size
Starts(2) is sends.size + receives.size
Starts(3) is sends.size + receives.size + news.size
...
Starts(5) is everything but filtered expressions, and so that's where we start counting.
Starts(6) is everything, so we use that to see if we can skip a split.